### DISCUSSION TITLE
How to access an array from an OPC-UA client

### DISCUSSION THREAD

**Discussion by SudhakarKuma** [FreeOpcUa/python-opcua#1304] [General]
Hello everyone! 

I am working on a **server-client** model for my project. In this project, I have a plant/ system running in **OpenModelica** (an open-source tool for simulating complex systems), which has a built-in **OPC-UA server**. I invoke this server for my model and then try to connect it with an **OPC-UA client**. For the client, I am making use of the **Python** **opcua** package.  

In my server-client model, I have an **array** of elements on the server. For my project, I need to access this array on the client-side. Accordingly, I need to modify some of its elements and write them back to the server. In the past, I have worked on accessing single values from the client. However, I am not able to figure out how to access an array using the **OPC-UA client**. Could you please help me with this? 

Regards,
Sudhakar

**Comment by AndreasHeine** [comment]
https://reference.opcfoundation.org/v104/Core/docs/Part6/5.1.6/

https://github.com/FreeOpcUa/python-opcua/blob/858274721766907148b527916e97446dbbf6af6d/opcua/ua/uatypes.py#L708

```python
from opcua import ua

# ua.Variant(value, varianttype)

ua.Variant([1,2,3], ua.VariantType.UInt16)
```

**Reply by SudhakarKuma** [reply]
@AndreasHeine Thanks for your response. Actually, I have an array as given below in OpenModelica: 

`
input Boolean valveProduct[2] = {false, false}; 
`

Now in my OPC-UA client, I want to access this array and change the values of its elements. Could you please suggest to me what Python OPC-UA command should I use in my client code for this?

**Reply by AndreasHeine** [reply]
should be simple like this:

https://github.com/FreeOpcUa/python-opcua/blob/858274721766907148b527916e97446dbbf6af6d/opcua/client/client.py#L658

```python
values = client.get_values([
        client.get_node("<nodeid>"),
])

# values should be a list of values in your case, so you simply access the list items via index
bool_list = values[0] # values is a list the first itmes containse the result for the <nodeid> above

# again just use the index to access it
bool_list[0] = False
bool_list[1] = True # modify a value

client.set_values([client.get_node("<nodeid>")], [bool_list]) # write the modified list back to the node
```

**Reply by AndreasHeine** [reply]
we also provide examples like: https://github.com/FreeOpcUa/python-opcua/blob/858274721766907148b527916e97446dbbf6af6d/examples/client-example.py

**Reply by SudhakarKuma** [reply]
@AndreasHeine  Thanks a lot for your prompt response. I tried writing the client the way you suggested. However, it is not working. So, I have an array in my server as given below:

` 
input Boolean valveProduct[2] = {false, true}; 
`

Now, I checked for the NodeId of this array in UaExpert. The NodeId is `ns=1;i=200000000` 

Next, I write the Python OPC-UA client as given below:

`
values = client.get_values([client.get_node("ns=1;i=200000000"),])
`
`
bool_list = values[0]
`

`
bool_list[0] = False
`
`
bool_list[1] = True 
`
`client.set_values([client.get_node("ns=1;i=200000000")], [bool_list])
`

However, it throws the error 
`TypeError: 'bool' object does not support item assignment`

Could you please help me rectify the issue?

**Reply by SudhakarKuma** [reply]
@AndreasHeine I tried to debug the client code. So, I modified the code a bit, as given below. In this code, the modified the **values[0]** to **values**. 


```Python
from opcua import Client
from opcua import ua
import time
import logging


# Define the URL on which the server is broadcasting
url = "opc.tcp://1xx.16x.xx.1xx:4841"

if __name__ == "__main__":
    client = Client(url)
    logging.basicConfig(level=logging.WARN)

    try:
        client.connect()
        print("Client Connected")

        run = client.get_node(ua.NodeId(10001, 0))
        run.set_value(True)
        
        values = client.get_values([client.get_node("ns=1;i=200000000"), client.get_node("ns=1;i=200000001")])
        bool_list = values 
        print(bool_list)
        print(type(bool_list))
              
        t = 0
        while True: 
            t = t + 1

            bool_list[0] = False
            bool_list[1] = True 
            client.set_values([client.get_node("ns=1;i=200000000"), client.get_node("ns=1;i=200000001")], [bool_list])

            time.sleep(1)
            print("="*40)
    
    except KeyboardInterrupt:
        print("Stopping sequence!")

    finally:
        print("Done!")
        client.disconnect()
```

Now, I am getting an error `IndexError: list index out of range`, pointing towards some issue in the line `client.set_values`. Could you please help me out once more?

**Reply by AndreasHeine** [reply]
> TypeError: 'bool' object does not support item assignment

It should work!

```
Python 3.7.3 (v3.7.3:ef4ec6ed12, Mar 25 2019, 21:26:53) [MSC v.1916 32 bit (Intel)] on win32
Type "help", "copyright", "credits" or "license()" for more information.
>>> value = [False, False]
>>> value[0] = True
>>> print(value)
[True, False]
>>> 
```

**Reply by AndreasHeine** [reply]
```python
from opcua import Client, ua

client = Client("opc.tcp://192.168.10.1:4840")

client.connect()

values = client.get_values([
    client.get_node('ns=3;s="Simulation"."bool"'),
    ])
print("get_values1:", values)

bool_array = values[0]
print("bool_array1: ", bool_array)

bool_array[2] = True
print("bool_array2: ", bool_array)

# client.set_values([client.get_node('ns=3;s="Simulation"."bool"')], [bool_array])

# some servers like it this way (like S7-1500) if they throw a uaerror related to timestamps
client.uaclient.set_attributes(
    [
        ua.NodeId.from_string('ns=3;s="Simulation"."bool"')
    ], 
    [
        ua.DataValue(
            ua.Variant(bool_array, ua.VariantType.Boolean)
        )
    ]
)

values = client.get_values([
    client.get_node('ns=3;s="Simulation"."bool"'),
    ])
print("get_values2:", values)

client.disconnect()
```

Microsoft Windows [Version 10.0.19042.928]
(c) Microsoft Corporation. Alle Rechte vorbehalten.

C:\Users\...\Desktop>C:/Users/HeineAn/AppData/Local/Programs/Python/Python37-32/python.exe c:/Users/.../Desktop/array_client.py
Requested session timeout to be 3600000ms, got 30000ms instead
get_values1: [[False, False, False, False, False, False]]
bool_array1:  [False, False, False, False, False, False]
bool_array2:  [False, False, True, False, False, False]
get_values2: [[False, False, True, False, False, False]]

C:\Users\...\Desktop>

**Reply by SudhakarKuma** [reply]
@AndreasHeine Thank you so much for your help. Could you please share your server code here? As my server is written in Modelica language, your server code would help me draw an analogy for my project.

**Reply by AndreasHeine** [reply]
https://www.prosysopc.com/products/opc-ua-simulation-server/

**Reply by AndreasHeine** [reply]
ns=6;s=BooleanArray

**Reply by SudhakarKuma** [reply]
@AndreasHeine Thanks a lot. It seems that you are using the online OPC UA server. I will download it and try it out. Meanwhile, I will like to draw your attention to my setup. My server is running in OpenModelica. The code running at the server is as given below:


```Modelica
model randTest
  input Boolean valveProduct[2] = {false, true}; 
  Boolean value, value1; 
equation
  value = valveProduct[1]; 
  value1 = valveProduct[2]; 
end randTest;
```

Now, I execute this code by invoking the OPC-UA server. To know the node IDs, I am using UaExpert, which shows the output as given below:

![Screenshot from 2021-04-21 00-18-42](https://user-images.githubusercontent.com/12606778/115448418-4fc2ce80-a237-11eb-96e9-51973a530e32.png)

You may observe that the two elements of my array are denoted as **valveProduct[1]** and **valveProduct[2]**. Now, I write my client code as given below:

```Python
# https://jckantor.github.io/CBE30338/04.01-Implementing_PID_Control_with_Python_Yield_Statement.html

from opcua import Client, ua

import time
import logging

client = Client("opc.tcp://192.168.0.3:4841")

client.connect()

run = client.get_node(ua.NodeId(10001, 0))
run.set_value(True)
print("Current state of run : {}".format(run.get_value()))

values = client.get_values([
    client.get_node('ns=1;s="valveProduct"'),
    ])
print("get_values1:", values)

bool_array = values[0]
print("bool_array1: ", bool_array)

bool_array[2] = True
print("bool_array2: ", bool_array)

client.set_values([client.get_node('ns=1;s="valveProduct"')], [bool_array])

# some servers like it this way (like S7-1500) if they throw a uaerror related to timestamps
# client.uaclient.set_attributes(
#     [
#         ua.NodeId.from_string('ns=3;s="Simulation"."bool"')
#     ], 
#     [
#         ua.DataValue(
#             ua.Variant(bool_array, ua.VariantType.Boolean)
#         )
#     ]
# )

values = client.get_values([
    client.get_node('ns=1;s="valveProduct"'),
    ])
print("get_values2:", values)

client.disconnect()
``` 
This client shows an error, as shown below:

![Screenshot from 2021-04-21 00-22-33](https://user-images.githubusercontent.com/12606778/115448772-b942dd00-a237-11eb-8c30-65bfbb04defa.png)

I do understand that my array is not getting read properly by the client. Could you please help me rectify it?

**Reply by AndreasHeine** [reply]
the error tells you everything its a python "TypeError" not python-opcua related error...

typically you should check if a result (bool_array) is None like you see in the prints... that can happenif for example a variable is not initialized (Variant.Value = Null) it could also be that your nodeid is wrong / not existend / typo -> result would also be None
your screenshot from uaexpert shows that the nodes have a numeric identifier so this: 'ns=1;s="valveProduct"' will never workyou have to stick to numeric nodeids!

If you want to learn more about opcua (its quite complex on the begining):
https://reference.opcfoundation.org/

i can only tell you that the client works with all servers i tested it without any issue! 
if you want to write reliable code you need either make typechecks or try/catch those errors but this is basc python programing for this is a python tutorial more suitable then asking all that ;)

**Reply by SudhakarKuma** [reply]
@AndreasHeine Thank you for your response. In my case, the two elements of the array are at two different IDs, like

- valveProduct[1] at ns=1;i=200000000
- valveProduct[2] at ns=1;i=200000001

Thus, I need to know how to get these two elements in a single go and then implement the client the way you suggested. I have tried to do this by creating a list as shown below:

```Python
values = client.get_values([
    client.get_node('ns=1;i=200000000'),client.get_node('ns=1;i=200000001'),
    ])
print("get_values1:", values)

bool_array = values
print("bool_array1: ", bool_array)

bool_array[1] = True
print("bool_array2: ", bool_array)

client.set_values([client.get_node('ns=1;i=200000000'),client.get_node('ns=1;i=200000001')], [bool_array])
```

However, it shows an OPC-UA error as shown below:

![Screenshot from 2021-04-21 01-22-54](https://user-images.githubusercontent.com/12606778/115455868-235f8000-a240-11eb-96a4-797d40041fc1.png)


Could you please help me out with this?

**Reply by AndreasHeine** [reply]
it is also a basic python thing you pass a list of two nodes and a list with only one value... for sure index out of range on the value list with just one item in it!

**Reply by AndreasHeine** [reply]
please read the docstring of set_values()

**Reply by SudhakarKuma** [reply]
> it is also a basic python thing you pass a list of two nodes and a list with only one value... for sure index out of range on the value list with just one item in it!

@AndreasHeine In the line of set_values, I understand that in the command given below:

```Python
client.set_values([client.get_node('ns=1;i=200000000'),client.get_node('ns=1;i=200000001')], [bool_array])
```
the **bool_array** is a list with single elements. That's why I change this line to 

```Python
client.set_values([client.get_node('ns=1;i=200000000'),client.get_node('ns=1;i=200000001')], [bool_array][0])
```
Then, in this case, I am getting an error `opcua.ua.uaerrors._auto.BadWriteNotSupported: "The server does not support writing the combination of value, status and timestamps provided."(BadWriteNotSupported)`. 

Could you please help me out with this?

**Reply by AndreasHeine** [reply]
i already did: above "client.uaclient.set_attributes"

**Reply by SudhakarKuma** [reply]
@AndreasHeine Thank you so much for your patience and support. I tried the following command 

```Python
    client.uaclient.set_attributes(
        [
            ua.NodeId.from_string('ns=1;i=200000000'),ua.NodeId.from_string('ns=1;i=200000001')
        ], 
        [
            ua.DataValue(
                ua.Variant(bool_array[0], ua.VariantType.Boolean)
            ), 
            ua.DataValue(
                ua.Variant(bool_array[1], ua.VariantType.Boolean)
            )
        ]
    )
```

This command does not throw any errors. Could you please have a look and confirm that this is the right way?

**Reply by AndreasHeine** [reply]
depends on what this is:
bool_array​[​0​]
bool_array​[​1]

but it looks right!

**Reply by SudhakarKuma** [reply]
It is the elements of the array which has to be written back to the server. Now, my complete client code looks like

```Python
# https://jckantor.github.io/CBE30338/04.01-Implementing_PID_Control_with_Python_Yield_Statement.html

from opcua import Client, ua

import time
import logging

client = Client("opc.tcp://192.168.0.3:4841")

client.connect()

run = client.get_node(ua.NodeId(10001, 0))
run.set_value(True)
print("Current state of run : {}".format(run.get_value()))

try:

    while True: 
        values = client.get_values([
            client.get_node('ns=1;i=200000000'),client.get_node('ns=1;i=200000001'),
            ])
        print("get_values1:", values)

        bool_array = values
        print("bool_array1: ", bool_array)

        bool_array[1] = True
        print("bool_array2: ", bool_array)

        client.uaclient.set_attributes(
            [
                ua.NodeId.from_string('ns=1;i=200000000'),ua.NodeId.from_string('ns=1;i=200000001')
            ], 
            [
                ua.DataValue(
                    ua.Variant(bool_array[0], ua.VariantType.Boolean)
                ), 
                ua.DataValue(
                    ua.Variant(bool_array[1], ua.VariantType.Boolean)
                )
            ]
        )

        values = client.get_values([
            client.get_node('ns=1;i=200000000'),client.get_node('ns=1;i=200000001'),
            ])
        print("get_values2:", values)
        time.sleep(1)
        print("="*10)
finally:
    client.disconnect()
```

Even though this code does not throw any error, the values are not being modified at the server, as shown below:

![Screenshot from 2021-04-21 02-13-55](https://user-images.githubusercontent.com/12606778/115461631-45103580-a247-11eb-80bc-a388875c8a7b.png)

I am sorry to bother you. But, do you mind having a look at it once?

**Reply by AndreasHeine** [reply]
currently you dont access an array, like in the title (its a little missleading...), just single boolean variables!
in yellow marked: the variable you want to write has to have CurrentWrite in order to be written by clients!

this is how an array variable looks like in opcua 
![Unbenannt](https://user-images.githubusercontent.com/56362817/115504991-19a63e80-a279-11eb-9546-425454b12336.PNG)

**Reply by SudhakarKuma** [reply]
@AndreasHeine Thank you for your response. I tried accessing the array directly as you suggested, 

```Python
values = client.get_values([
    client.get_node('ns=3;s="Simulation"."bool"'),
    ])
print("get_values1:", values)

bool_array = values[0]
print("bool_array1: ", bool_array)
```
However, I had difficulty in doing so. So, I decided to access the array in the form of elements. However, it does not seem to be working. In your previous reply, you have attached a screenshot. May I know from which tool is this screenshot taken? 
Also, do you mind sharing the code of your server? I cannot download the server from the [link](https://www.prosysopc.com/products/opc-ua-simulation-server/) which you provided.

**Reply by AndreasHeine** [reply]
UaExpert
https://www.unified-automation.com/products/development-tools/uaexpert.html

i cannot share the servercode its a siemens plc... and for the prosys server you need to make a download request (free of charge) you only must register!

**Reply by SudhakarKuma** [reply]
@AndreasHeine Thank you for your response. I have been using UaExpert for my previous experiments. While accessing the array, **AccessLevel** and **UserAccessLevel** are shown as **None**.  However, even for other experiments where I have successfully communicated between client and server, **AccessLevel** and **UserAccessLevel** are shown as **None** for all the objects. 

Could you please educate me on why UaExpert is not showing the correct values of **AccessLevel** and **UserAccessLevel**?

**Reply by AndreasHeine** [reply]
please make a screenshot (from uaexpert) form your variable (like i did yesterday) and post it here!

**Reply by SudhakarKuma** [reply]
@AndreasHeine Thanks! I have attached the screenshot here. The variable which is being highlighted here is **tank1_outCtr**.  UaExpert shows its **AccessLevel** and **UserAccessLevel** as None, as shown below.  
![Screenshot from 2021-04-22 15-24-39](https://user-images.githubusercontent.com/12606778/115695126-22813800-a37f-11eb-8dfd-ad56fb5a897b.png)

However, I have been able to read and write this variable using my Python OPC-UA client code. Could you please have a look and help me with UaExpert?

**Reply by AndreasHeine** [reply]
this looks like a wrong opcua implementation in the server...

WriteMask = 1 -> only the first bit (0) is True means it is only readabel (AccessLevelType CurrentRead)

References:
https://reference.opcfoundation.org/v104/Core/DataTypes/AccessLevelType/
https://reference.opcfoundation.org/v104/Core/docs/Part3/5.6.2/

if a Variable is Writeable aswell it needs a WriteMask = 3 -> first (0) and second (1) bit is True (AccessLevelType CurrentRead, CurrentWrite)

if you are able to write the variable on the screenshot its not correct implemented!
what server is it? which opcua stack is used?

**Reply by SudhakarKuma** [reply]
@AndreasHeine Thanks! I am using the built-in OPC-UA server of OpenModelica. So, the OpenModelica compiler has an implementation of OPC UA using an [open source OPC UA implementation](https://open62541.org/). To invoke the OPC-UA server, I simulate the model with simulation flags like **-embeddedServer=opc-ua**. Accordingly, I connect my Python OPC-UA client with this server.

**Reply by AndreasHeine** [reply]
tbh the problem is not on python-opcua client side, it maybe better to ask that on there modellica repo!?

**Reply by SudhakarKuma** [reply]
Yeah. @AndreasHeine Thanks a lot. I will ask it on the Modelica repo.

**Reply by SudhakarKuma** [reply]
@AndreasHeine Hi! As suggested, I am working with Prosys tools. So, I am using **Prosys OPC UA Simulation Server** and **Prosys OPC UA Monitor**. We can use **Simulation Server** to create objects which can be simulated on the server.  Next, we can use **Monitor** to read those objects. I have tried connecting the **Simulation Server** and  **Monitor** with base data types. Now, I wish to create an object which is an array (for example, [1, 2.5, 9.0] or [True, False]) in the  **Simulation Server**. Accordingly, I plan to read this array and modify its elements. Could you please educate me on how to simulate an array in **Simulation Server**? Please note that I wish to modify the elements of this array by using a client. 

Also, I believe that **Monitor** cannot write values on the server. So, do you mind suggesting to me some Prosys tools that can modify the server's values? I came across **Prosys OPC UA Browser**, which might be helpful. Could you please confirm it once?

**Reply by AndreasHeine** [reply]
unfortunately not! if you have trouble with a prosys product you should them for help. we can only provide support if there are issues between or check complience to spec. in those cases.

**Reply by SudhakarKuma** [reply]
> ```python
> from opcua import Client, ua
> 
> client = Client("opc.tcp://192.168.10.1:4840")
> 
> client.connect()
> 
> values = client.get_values([
>     client.get_node('ns=3;s="Simulation"."bool"'),
>     ])
> print("get_values1:", values)
> 
> bool_array = values[0]
> print("bool_array1: ", bool_array)
> 
> bool_array[2] = True
> print("bool_array2: ", bool_array)
> 
> # client.set_values([client.get_node('ns=3;s="Simulation"."bool"')], [bool_array])
> 
> # some servers like it this way (like S7-1500) if they throw a uaerror related to timestamps
> client.uaclient.set_attributes(
>     [
>         ua.NodeId.from_string('ns=3;s="Simulation"."bool"')
>     ], 
>     [
>         ua.DataValue(
>             ua.Variant(bool_array, ua.VariantType.Boolean)
>         )
>     ]
> )
> 
> values = client.get_values([
>     client.get_node('ns=3;s="Simulation"."bool"'),
>     ])
> print("get_values2:", values)
> 
> client.disconnect()
> ```
> 
> Microsoft Windows [Version 10.0.19042.928]
> (c) Microsoft Corporation. Alle Rechte vorbehalten.
> 
> C:\Users...\Desktop>C:/Users/HeineAn/AppData/Local/Programs/Python/Python37-32/python.exe c:/Users/.../Desktop/array_client.py
> Requested session timeout to be 3600000ms, got 30000ms instead
> get_values1: [[False, False, False, False, False, False]]
> bool_array1: [False, False, False, False, False, False]
> bool_array2: [False, False, True, False, False, False]
> get_values2: [[False, False, True, False, False, False]]
> 
> C:\Users...\Desktop>

@AndreasHeine Thanks for your response. Which simulation server did you use for this?

**Reply by AndreasHeine** [reply]
i use real systems and hardware for testing!

**Reply by SudhakarKuma** [reply]
@AndreasHeine Could you please elaborate a bit? You might remember that I have been working with OpenModelica, which has a built-in OPC UA server. However, this server has some issues like the inability to write Boolean values, not having a single node for an array, etc. So, I wish to check how other OPC UA server handles these issues.  Could you please suggest some other tools with a built-in OPC UA server, as OpenModelica has?

**Reply by AndreasHeine** [reply]
just google "opc ua simulation server" or "opc ua demo server" ;)

**Reply by SudhakarKuma** [reply]
Hi, @AndreasHeine Thanks for your response and constant support. I have come across a few OPC UA simulation servers and will try these out shortly. Meanwhile. I thought of using the **Server** module of the Python OPC UA package. First, I wanted to see how an array is handled and how to modify the array elements. So, I  have a server (having an array `[1, 2, 6]`) as given below:

```Python
from opcua import Server
import time

server = Server()

url = "opc.tcp://192.168.0.3:4840"
server.set_endpoint(url)

name = "OPCUA_Simulation_Server"
addspace = server.register_namespace(name)

node = server.get_objects_node()
params = node.add_object(addspace, "Parameters")

arr = params.add_variable(addspace, "myArray", [1, 2, 6])
arr.set_writable()

server.start()
print("Server started at {}".format(url))

while True:
    print("Array is: {}".format(arr.get_value()))
    time.sleep(1)
```

Now, suppose we wish to have a client that can modify the array to  `[1, 5.6, 6]`. To do so, one way to write the client code is as given below:

```Python
from opcua import Client
import time

url = "opc.tcp://192.168.0.3:4840"

if __name__ == "__main__":
    client = Client(url)

    try:
        client.connect()
        print("Client connected!")

        while True:
            arr = client.get_node("ns = 2; i = 2")
            arrVal = arr.get_value()
            print(arrVal)

            arr.set_value([1, 5.6, 6])

            time.sleep(1)

    except KeyboardInterrupt:
        print("Stopping Sequence!")
    
    finally:
        print("Done!")
        client.disconnect()
 ```
Could you please suggest some other ways, which would help to modify the values at a particular index (or a group of indices)? It would eliminate the need to pass all the values in the `set_value` method.

**Reply by AndreasHeine** [reply]
https://reference.opcfoundation.org/v104/Core/docs/Part4/5.10.4/
the feature you are looking for is called "indexrange" in the write parameters, but i dont think we support that in python-opcua!

or do we? @oroulet

**Reply by GrantEdwards** [reply]
Does python-opcua support indexed writes?

**Reply by AndreasHeine** [reply]
> Does python-opcua support indexed writes?

it is better to open own issues or discussions   rather then post on old ones with a different topic!

if you use the lowlevel-client you can set the request params manually (if the server supports that feature)

**Comment by SudhakarKuma** [comment]
Hi @AndreasHeine With your advice, I have been able to work with 1D arrays on the OPC-UA server. Now, I am facing issues with models having 2D arrays. When I connect to these models using a Python OPC-UA client, it works fine during the simulation. However, at the end of the simulation, I get an array of **invalid pointer** as shown below: 

![Screenshot from 2021-05-18 14-19-14](https://user-images.githubusercontent.com/12606778/118621698-3edc8d00-b7e4-11eb-93b1-723421e55356.png)

The same model works perfectly fine without an OPC-UA server. But when I simulate it on the OPC-UA server, I get this error, as shown above. That's why it makes me believe that the simulation has something to do with OPC-UA only. Could you please provide me with some insights into this?

**Reply by swamper123** [reply]
> I get an array of invalid pointer as shown below [...]`free(): invalid pointer`

Well this neither looks like a python error nor a Client error. The opc-ua server should tell you what is wrong and should throw typical error messages like `BadAttributeError` etc. Otherwise the server is not stable.

**Reply by SudhakarKuma** [reply]
@swamper123 Thank you for your prompt response. On the client side, the exchange of data happens properly. So, should I assume that my OPC-UA server is having some issues?

**Reply by swamper123** [reply]
If a request is faulty, the server should respond properly with a [common status code](https://reference.opcfoundation.org/v104/Core/docs/Part4/7.34.2/) and not crash, like this. 
And if you say yourself, that the Client is doing his job propperly, it has to be the server.
